onnxruntime/core/optimizer/transpose_optimizer/api_impl.cc

```diff
@@ -562,60 +576,128 @@ void ApiGraph::ReshapeInitializer(std::string_view name, const std::vector<int64
   }
 
+  node.SetExecutionProviderType(std::string(node_ep));
+
   for (size_t i = 0; i < input_args.size(); ++i) {
     NodeArg* arg = input_args[i];
     if (arg->Exists()) {
       const std::string& name_str = arg->Name();
-      graph_.AddConsumerNode(name_str, &node);
-      const auto* inp_node = graph_.GetProducerNode(name_str);
+      graph.AddConsumerNode(name_str, &node);
+      const auto* inp_node = graph.GetProducerNode(name_str);
       if (inp_node != nullptr) {
         int inp_node_out_index = graph_utils::GetNodeOutputIndexFromOutputName(*inp_node, name_str);
-        graph_.AddEdge(inp_node->Index(), node.Index(), inp_node_out_index, gsl::narrow_cast<int>(i));
+        graph.AddEdge(inp_node->Index(), node.Index(), inp_node_out_index, gsl::narrow_cast<int>(i));
       }
     }
   }
 
   for (NodeArg* arg : output_args) {
-    graph_.UpdateProducerNode(arg->Name(), node.Index());
+    graph.UpdateProducerNode(arg->Name(), node.Index());
+  }
+
+  return node;
+}
+
+#if defined(ORT_EXTENDED_MINIMAL_BUILD)
+// This is a list of onnx ops and their versions which transpose_optimizer can potentially add to the graph.
+// This is needed in minimal build since opschema is not available.
+// The versions should be sorted.
+static const std::unordered_map<std::string, std::vector<int>> onnx_ops_available_versions = {
+    {"Squeeze", {1, 11, 13}},
+    {"Unsqueeze", {1, 11, 13}},
+    {"Gather", {1, 11, 13}},
+    {"Transpose", {1, 13}},
+    {"Identity", {1, 13, 14, 16}},
+};
+
+// Based on the opset version imported for this model, returns the since version for the node.
+static int GetSinceVersionForNewOp(std::string_view op_type, std::string_view domain,
+                                   const std::unordered_map<std::string, int>& domain_to_version_map) {
+  int since_version = -1;
+  auto opset_import_iter = domain_to_version_map.find(std::string(domain));
+  if (opset_import_iter != domain_to_version_map.end()) {
+    int opset_version = opset_import_iter->second;
+    auto iter = onnx_ops_available_versions.find(std::string(op_type));
+    ORT_ENFORCE(iter != onnx_ops_available_versions.end(),
+                "Transpose Optimizer is adding an unexpected node: ", op_type,
+                "An entry for this node should be added in onnx_ops_available_versions and static_kernel_hashes map.");
+
+    for (auto version : iter->second) {
+      if (version <= opset_version) {
+        since_version = version;
+      }
+    }
   }
 
+  return since_version;
+}
+#endif
+
+std::unique_ptr<api::NodeRef> ApiGraph::AddNode(std::string_view op_type,
+                                                const std::vector<std::string_view>& inputs, size_t num_outputs,
+                                                std::string_view domain) {
+  int since_version =
+#if defined(ORT_EXTENDED_MINIMAL_BUILD)
+      GetSinceVersionForNewOp(op_type, domain, graph_.DomainToVersionMap());
+#else
+      -1;
+#endif
```

Comment:
Do we need different logic for a minimal build (i.e. would that SinceVersion be wrong if it wasn't a minimal build)?

Reply:
In case of full build we use SetOpSchemaFromRegistry for fetching the schema as well as the since version for this newly added op. Maybe we can just use this... needs testing